routing/matcher.go

```diff
@@ -1,111 +1,335 @@
+
+// A Matcher represents a preprocessed set of definitions and their associated
+// values.
+type matcher struct {
+	paths               *pathmux.Tree
+	rootLeaves          leafMatchers
+	ignoreTrailingSlash bool
 }
 
-func (r *Routing) createFilter(def *eskip.Filter) (filters.Filter, error) {
-	spec, ok := r.filterRegistry[def.Name]
-	if !ok {
-		return nil, fmt.Errorf("filter not found: '%s'", def.Name)
+// An error created if a definition cannot be preprocessed.
+type definitionError struct {
+	Id       string
+	Index    int
+	Original error
+}
+
+func (err *definitionError) Error() string {
+	if err.Index < 0 {
+		return err.Original.Error()
 	}
 
-	return spec.CreateFilter(def.Args)
+	return fmt.Sprintf("%s [%d]: %v", err.Id, err.Index, err.Original)
+}
+
+var freeWildcardRx *regexp.Regexp
+
+func init() {
+	freeWildcardRx = regexp.MustCompile("/[*][^/]+$")
 }
 
-func (r *Routing) createFilters(def *eskip.Route) ([]filters.Filter, error) {
-	fs := make([]filters.Filter, len(def.Filters))
-	for i, fdef := range def.Filters {
-		f, err := r.createFilter(fdef)
+func compileRxs(exps []string) ([]*regexp.Regexp, error) {
+	rxs := make([]*regexp.Regexp, len(exps))
+	for i, exp := range exps {
+		rx, err := regexp.Compile(exp)
 		if err != nil {
 			return nil, err
 		}
 
-		fs[i] = f
+		rxs[i] = rx
 	}
 
-	return fs, nil
+	return rxs, nil
 }
 
-func (r *Routing) convertDef(def *eskip.Route) (*routeDef, error) {
-	b, err := createBackend(def)
+func newLeaf(r *Route) (*leafMatcher, error) {
+	hostRxs, err := compileRxs(r.HostRegexps)
 	if err != nil {
 		return nil, err
 	}
 
-	fs, err := r.createFilters(def)
+	pathRxs, err := compileRxs(r.PathRegexps)
 	if err != nil {
 		return nil, err
 	}
 
-	rt := &Route{b, fs}
-	return &routeDef{def, rt}, nil
+	headerExps := r.HeaderRegexps
+	allHeaderRxs := make(map[string][]*regexp.Regexp)
+	for k, exps := range headerExps {
+		headerRxs, err := compileRxs(exps)
+		if err != nil {
+			return nil, err
+		}
+
+		allHeaderRxs[k] = headerRxs
+	}
+
+	return &leafMatcher{
+		method:        r.Method,
+		hostRxs:       hostRxs,
+		pathRxs:       pathRxs,
+		headersExact:  r.Headers,
+		headersRegexp: allHeaderRxs,
+		route:         r}, nil
 }
 
-func (r *Routing) convertDefs(eskipDefs []*eskip.Route) []requestmatch.Definition {
-	matcherDefs := []requestmatch.Definition{}
-	for _, d := range eskipDefs {
-		rd, err := r.convertDef(d)
-		if err == nil {
-			matcherDefs = append(matcherDefs, rd)
-		} else {
-			// idividual definition errors are accepted here
-			log.Println(err)
+func freeWildcardParam(path string) string {
+	param := freeWildcardRx.FindString(path)
+	if param == "" {
+		return ""
+	}
+
+	// clip '/*' and return only the name
+	return param[2:]
+}
+
+// Constructs a Matcher based on the provided definitions. If `ignoreTrailingSlash`
+// is true, the Matcher handles paths with or without a trailing slash equally.
+func newMatcher(rs []*Route, ignoreTrailingSlash bool) (*matcher, []*definitionError) {
```

Comment:
boolean flags in APIs are a smell to me. I'd prefer a set of Options (enums), that currently only have one value `ignoreTrailingSlashes`. Either provide a singletonSet with that value or an empty set to change the behaviour.